### PR TITLE
Convert File Uploads to Multi-Part Form

### LINKED ISSUE #927: API Upload Problems[BUG]
I followed all the steps from the README to automate file uploads, but nothing works. I keep getting an error message:

`{"statusCode":401, "message":"Unauthorized"}`

Curl commands:
 - Generate access token:
`curl -X POST -H "Content-Type: application/json" -d '{"email": "user@example.com", "password": "password" }' http://localhost:3000/authn/login`

 - Attempt to upload
`curl -X POST -H "Content-Type: application/json" -H 'Authorization: Bearer  {"userID":"2", "accessToken":"<long access token string>"}' -d "@/home/vcscan/scans/02102021/<filename>.json" http://localhost:3000/evaluations`

I tried different variations of the upload command, omitting the "Authorization" or "Bearer" words, just using the token json, etc. and it either provides no output (with no upload), or it gives the error message above.

### PR DESCRIPTION
**Pull request by camdenmoors**
Closes #927, Changes the upload evaluation endpoint to be a multi-part file upload to allow for uploads via the command line.

For example:
```
curl -F "data=@/data/evaluations/nginx_sample.json" \
	-F "filename=Nginx Sample Clean" \
	-F "public=true" \
	-H "Authorization: Bearer yourbearer" \
	"http://localhost:3000/evaluations"
```

### CODE REVIEW

File: apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue [hunk 1]
```diff
@@ -91,15 +91,16 @@ export default class FileItem extends mixins(ServerMixin) {
...
-    let evaluationDTO: ICreateEvaluation = {
-      data: file.evaluation.data,
-      filename: file.filename,
-      userId: ServerModule.userInfo.id,
-      evaluationTags: []
-    };
+    const formData = new FormData();
+    formData.append("filename", file.filename);
+    var blob = new Blob([JSON.stringify(file.evaluation.data)], {type: 'text/plain'});
+    console.log(blob);
```
Comment (rbclark) [inline apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue:97]: This should be removed.

File: apps/backend/src/evaluations/dto/create-evaluation.dto.ts [hunk 2]
```diff
@@ -1,26 +1,17 @@
...
 
-  @IsNotEmpty()
-  @IsObject()
   readonly data!: Record<string, any>;
 
   @IsNotEmpty()
-  @IsBoolean()
-  readonly public!: boolean;
+  @IsIn([true, false, 'true', 'false'])
+  public!: boolean | string;
```
Comment (rbclark) [inline apps/backend/src/evaluations/dto/create-evaluation.dto.ts:14]: Can the DTO constructor handle this conversion instead of allowing strings here>

File: README.md [hunk 3]
```diff
@@ -163,7 +163,7 @@ curl -X POST -H "Content-Type: application/json" -d '{"email": "user@example.com
 curl -X POST -H "Content-Type: application/json" -d '{"email": "user@example.com", "password": "password" }' http://localhost:3000/authn/login
 # The previous command returns a Bearer Token that needs to get placed in the following command
 # Upload evaluation
-curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer bearertokengoeshere" -d "@/home/example/Desktop/test.json" http://localhost:8000/evaluations
+curl -F "data=@/Full/Path/To/Your/Evaluation.json" -F "filename=Your Filename" -F "public=true/false" -H "Authorization: Bearer bearertokengoeshere" "http://localhost:3000/evaluations"
```
Comment (rbclark) [inline README.md:166]: Not sure if it is necessary to specify that the JSON is a `/full/path`, I believe relative paths work here as well.

File: apps/backend/src/evaluations/dto/create-evaluation.dto.ts [hunk 4]
```diff
@@ -1,26 +1,16 @@
...
 import {CreateEvaluationTagDto} from '../../evaluation-tags/dto/create-evaluation-tag.dto';
 
 export class CreateEvaluationDto implements ICreateEvaluation {
   @IsNotEmpty()
   @IsString()
   readonly filename!: string;
 
-  @IsNotEmpty()
-  @IsObject()
   readonly data!: Record<string, any>;
```
Comment (rbclark) [inline apps/backend/src/evaluations/dto/create-evaluation.dto.ts:10]: Is this field applicable at all to the CreateEvaluationDto at this point?

File: apps/backend/src/evaluations/dto/create-evaluation.dto.ts [hunk 5]
```diff
@@ -1,26 +1,16 @@
...
 
   @IsNotEmpty()
-  @IsBoolean()
-  readonly public!: boolean;
+  public!: boolean;
```
Comment (rbclark) [inline apps/backend/src/evaluations/dto/create-evaluation.dto.ts:13]: Shouldn't `IsBoolean` still be present here?

File: apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue (hunk 1 above)
Comment (rbclark) [reply apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue:95]: The console.log is still present.

File: apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue [hunk 6]
```diff
@@ -91,15 +91,16 @@ export default class FileItem extends mixins(ServerMixin) {
...
+    formData.append("data", blob);
+    // Seems that publicty hasn't been setup yet
```
Comment (rbclark) [inline apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue:97]: Is this relevant to include here? This is being handled as part of the parent PR this is referencing.

File: libs/interfaces/evaluation/create-evaluation.interface.ts [hunk 7]
```diff
@@ -3,6 +3,6 @@ import {ICreateEvaluationTag} from '..';
 export interface ICreateEvaluation {
   readonly filename: string;
   readonly data: Record<string, any>;
-  readonly public: boolean;
+  public: boolean | string;
```
Comment (rbclark) [inline libs/interfaces/evaluation/create-evaluation.interface.ts:6]: This should still be just boolean.

**Review by rbclark — CHANGES_REQUESTED** [review]

File: apps/backend/src/evaluations/dto/create-evaluation.dto.ts (hunk 5 above)
Comment (camdenmoors) [reply apps/backend/src/evaluations/dto/create-evaluation.dto.ts:13]: It doesn't work with IsBoolean, it still returns an API error if the field is a string.

File: apps/backend/src/evaluations/dto/create-evaluation.dto.ts (hunk 4 above)
Comment (camdenmoors) [reply apps/backend/src/evaluations/dto/create-evaluation.dto.ts:10]: Data is still needed as part of CreateEvaluationDto as the controller converts the blob to the data value on a new CreateEvaluationDto

File: apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue (hunk 6 above)
Comment (camdenmoors) [reply apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue:97]: This is just here to make it work, I can remove the comment though

File: apps/backend/src/evaluations/evaluations.controller.ts [hunk 8]
```diff
@@ -53,13 +57,25 @@ export class EvaluationsController {
...
+    createEvaluationDto: CreateEvaluationDto,
+    @UploadedFile() data: Express.Multer.File,
     @Request() request: {user: User}
   ): Promise<EvaluationDto> {
+    const updatedEvaluationDto: CreateEvaluationDto = {
+      filename: createEvaluationDto.filename,
+      evaluationTags: createEvaluationDto.evaluationTags || [],
+      data: JSON.parse(data.buffer.toString('utf8')),
+      public: createEvaluationDto.public
+    };
```
Comment (rbclark) [inline apps/backend/src/evaluations/evaluations.controller.ts:72]: Shouldn't this just be

```javascript
    const updatedEvaluationDto: CreateEvaluationDto = {
      ...createEvaluationDto,
      data: JSON.parse(data.buffer.toString('utf8')),
    };
```

File: apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue [hunk 9]
```diff
@@ -91,15 +89,14 @@ export default class FileItem extends mixins(ServerMixin) {
   save_evaluation(file: EvaluationFile) {
     this.saving = true;
 
-    let evaluationDTO: ICreateEvaluation = {
-      data: file.evaluation.data,
-      filename: file.filename,
-      userId: ServerModule.userInfo.id,
-      evaluationTags: []
-    };
```
Comment (rbclark) [inline apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue:99]: Does this remove the ability to create tags?

Comment (rbclark) [reply apps/frontend/src/components/global/sidebaritems/SidebarFileList.vue:99]: The tag functionality shouldn't need to live in inspecjs at all, for the way heimdall2 is setup it doesn't really make sense to add tags on save, however for someone uploading from their own CI job they would probably want the ability to add tags and groups as part of the save.

**Review by rbclark — APPROVED** [review]
